compiler_opt/rl/problem_configuration.py

```diff
@@ -100,6 +105,143 @@ def get_nonnormalized_features(self) -> Iterable[str]:
   def get_runner_type(self) -> 'type[compilation_runner.CompilationRunner]':
     raise NotImplementedError
 
+  @abc.abstractmethod
+  def get_module_specs(
+      self,
+      data_path: str,
+      additional_flags: Tuple[str, ...] = (),
+      delete_flags: Tuple[str, ...] = ()
+  ) -> List[ModuleSpec]:
+    """Fetch a list of ModuleSpecs for the corpus at data_path
+
+    Args:
+      data_path: base directory of corpus
+      additional_flags: tuple of clang flags to add.
+      delete_flags: tuple of clang flags to remove.
+    """
+    raise NotImplementedError
+
+  @staticmethod
+  def _get_module_specs(data_path: str, additional_flags, delete_flags,
+                        xopts) -> List[ModuleSpec]:
+    module_paths: List[str] = get_module_paths(data_path)
+
+    is_thinlto: bool = is_thinlto_fn(module_paths)
+    has_cmd: bool = has_cmd_fn(module_paths)
+
+    # TODO: (b/233935329) Per-corpus *fdo profile paths can be read into
+    # {additional|delete}_flags here
+    meta = load_metadata(data_path)
+
+    xopts.update({'output': ['-o', '{path:s}']})
+
+    module_specs: List[ModuleSpec] = []
 
-def is_thinlto(module_paths: Iterable[str]) -> bool:
+    # Future: cmd_override could have per-module overrides if needed
+    cmd_override = get_cmd_override(meta, critical=not has_cmd)
+
+    for module_path in module_paths:
+      exec_cmd = get_command_line_for_bundle(
+          cmd_file=(module_path + '.cmd') if has_cmd else None,
+          ir_file=module_path + '.bc',
+          thinlto_file=(module_path + '.thinlto.bc') if is_thinlto else None,
+          additional_flags=additional_flags,
+          delete_flags=delete_flags,
+          cmd_override=list(cmd_override) if cmd_override is not None else None)
+      module_specs.append(ModuleSpec(exec_cmd, xopts, module_path))
+
+    return module_specs
+
+
+def is_thinlto_fn(module_paths: Iterable[str]) -> bool:
   return tf.io.gfile.exists(next(iter(module_paths)) + '.thinlto.bc')
+
+
+def has_cmd_fn(module_paths: Iterable[str]) -> bool:
+  return tf.io.gfile.exists(next(iter(module_paths)) + '.cmd')
+
+
+def get_module_paths(data_path: str) -> List[str]:
+  with open(
+      os.path.join(data_path, 'module_paths'), 'r', encoding='utf-8') as f:
+    return [os.path.join(data_path, name.rstrip('\n')) for name in f]
+
+
+def load_metadata(data_path: str) -> Dict[any, any]:
+  try:
+    with open(
+        os.path.join(data_path, 'metadata.json'), 'r', encoding='utf-8') as f:
+      return json.load(f)
+  except FileNotFoundError:
+    logging.info('metadata.json couldn\'t be found.')
+    return {}
+
+
+def get_cmd_override(meta: Dict[any, any],
+                     critical=False) -> Union[List[str], None]:
+  if 'global_command_override' in meta:
```

Comment:
what's the use case for global_command_override?

Reply:
it'll be for training on corpora extracted from the linker (in non-distributed ThinLTO cases); since we're having clang invoke ThinLTO in the training cycles instead, we can't reuse the original command as-is. Currently I'm still missing `ir_extractor.py` changes to make the workflow easier, so I haven't wrote instructions for release on GitHub yet. 

There's instructions in my internal design doc to get it running for now though